keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1116 @@
+  determine whether a connection matching a particular rule/policy must be
+  logged or not.
+
+## Design Details
+
+### AdminNetworkPolicy API Design
+
+The following new `AdminNetworkPolicy` API will be added to the `policy.networking.k8s.io` 
+API group.  
+
+```golang
+type AdminNetworkPolicy struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+  // Specification of the desired behavior of AdminNetworkPolicy.
+  Spec     AdminNetworkPolicySpec
+  // Status to be reported by the CNI, this is not standardized in alpha and 
+  // CNIs should report what they see fit in relation to their AdminNetworkPolicy
+  // implementation
+  // +optional 
+  Status   []metav1.Condition
+}
+
+type AdminNetworkPolicySpec struct {
+  // Int value bound to 0 - 999, with 0 signaling the special "baseline" priority
+  Priority      Int 
+  // Allows for multiple AdminNetworkPolicy rules at a single priority level
+  // If empty no rules are applied
+  // A total of 100 total rules will be allowed per each network policy instance, 
+  // this rule count will be calculated as the total summation of the 
+  // AdminNetworkPolicyRules, AdminNetworkPolicyIngressRules, and AdminNetworkPolicyEgressRules
+  // in a single AdminNetworkPolicy Instance. 
+  PolicyRules   []AdminNetworkPolicyRule
+}
+
+type AdminNetworkPolicyRule struct {
+  // No implicit isolation of AppliedTo Pods/Namespaces. 
+  // If no rules are a defined no behavior will be observed 
+  AppliedTo   AppliedTo
+  Ingress     []AdminNetworkPolicyIngressRule
+  Egress      []AdminNetworkPolicyEgressRule
+}
+
+// Exactly one of the selector pointers should be set in a given AppliedTo
+type AppliedTo struct {
+  NamespaceSelector         *metav1.LabelSelector
+  NamespaceAndPodSelector   *NamespaceAndPodSelector
+}
+
+// NamespaceAndPodSelector allows the user to select a namespace/s and optionally 
+// select a given set of pods in that namespace/s
+type NamespaceAndPodSelector struct {
+  NamespaceSelector   metav1.LabelSelector
+  // Used to explicitly select pods within a namespace
+  PodSelector         *metav1.LabelSelector
+}
+
+type AdminNetworkPolicyIngressRule struct {
+  // Name is an identifier for this rule. Must be unique within an instance of ANP. Can be
+  // same across policy instances.
+  // +optional
+  Name      string
+  // Action specifies whether this rule must pass, allow or deny traffic. 
+  // Allow: explicitly allows the selected traffic 
+  // Deny: explicitly denies the selected traffic
+  // Pass: explicitly allows the selected traffic to skip and remaining ANP rules 
+  // and be delegated by K8's Network Policy. 
+  Action    RuleAction
+  // List of ports for incoming traffic.
+  // Each item in this list is combined using a logical OR. If this field is
+  // empty or missing, this rule matches no ports (traffic not restricted by port).
+  // If this field is present and contains at least one item, then this rule 
```

Comment:
Let's make this reaally explicit, something like:

// The destination ports allowed for incoming traffic.  Items in this list are combined using a logical OR - the rule 
// applies to traffic which matches at least one port in the list.  If this field has no elements, the default value is to 
// match all ports.

Now the question is - is that the semantic we want.  Is it better to be explicit (no value here means no ports) or implicit?  I sort of think NetPol was backwards and we should err on  the side of explicit for security.  That said, it would be different from NetPol.

Reply:
So as I specified above the goal here was to flip the semantics, now an empty `Ports` array means no ports, and if the user want's to explicitly reference all TCP ports, they can define an empty `AdminNetworkPolicyPort` like so 

```
to: 
- NamespaceSetSelector: 
     scope: Selector
     namespaceSelector 
  Ports: 
     protocol: TCP
     Port: {} 
```

But I guess that also means I need to update all my sample yamls...

Reply:
> But I guess that also means I need to update all my sample yamls...

If the majority of the user stories don't care about specific ports, then that suggests that the default behavior should be to not care about specific ports.

Reply:
Dan, are you OK with `allPorts` as an explicit way of expressing this?